extensions/android-paging3/src/main/java/com/squareup/sqldelight/android/paging3/OffsetQueryPagingSource.kt

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.sqldelight.android.paging3
+
+import androidx.paging.PagingState
+import com.squareup.sqldelight.Query
+import com.squareup.sqldelight.Transacter
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+
+internal class OffsetQueryPagingSource<RowType : Any>(
+  private val queryProvider: (limit: Long, offset: Long) -> Query<RowType>,
+  private val countQuery: Query<Long>,
+  private val transacter: Transacter,
+  private val dispatcher: CoroutineDispatcher,
+) : QueryPagingSource<Int, RowType>() {
+
+  override suspend fun load(
+    params: LoadParams<Int>
+  ): LoadResult<Int, RowType> = withContext(dispatcher) {
+    try {
+      val key = params.key?.toLong() ?: 0L
+      transacter.transactionWithResult {
+        val count = countQuery.executeAsOne()
+        if (isOutOfBounds(key, params.loadSize, count)) throw IndexOutOfBoundsException()
+
+        val data = queryProvider(params.loadSize.toLong(), params.loadSize * key)
+          .also { currentQuery = it }
+          .executeAsList()
+
+        LoadResult.Page(
+          data = data,
+          prevKey = if (params.key == 0) null else params.key?.dec(),
+          nextKey = if (isOutOfBounds(key.inc(), params.loadSize, count)) null else key.inc().toInt()
+        )
+      }
+    } catch (e: Exception) {
```

Comment:
Blanket `catch (e: Exception)` will swallow `IndexOutOfBoundsException` and pass it to rest of paging as a recoverable error / make it visible to presenter-layer for retry, etc - is that intended? Are there `IOException` type errors SqlDelight would throw that you wouldn't want to catch here as well?

Reply:
That's a good point. While writing I did think that this should be caught. Your mentioning of "recoverable" makes me think otherwise. Good catch.

Reply:
Done!